src/SASViyaApiClient.ts

```diff
@@ -434,15 +491,96 @@ export class SASViyaApiClient {
     const formData = new NodeFormData()
     formData.append('file', contentBuffer, fileName)
 
-    return (
-      await this.requestClient.post<File>(
-        `/files/files?parentFolderUri=${parentFolderUri}&typeDefName=file#rawUpload`,
-        formData,
-        accessToken,
-        'multipart/form-data; boundary=' + (formData as any)._boundary,
-        headers
-      )
-    ).result
+    /** Query Viya for file metadata based on extension type. */
+
+    // typeDefName - Viya accepts this property during the file creation
+    let typeDefName: string | undefined = undefined
+    // Additional properties are supplied by a patch.
+    let filePatch:
+      | {
+          name: string
+          properties: IViyaTypesProperties | undefined
+        }
+      | undefined = undefined
+
+    // The patching process requires properties related to the file-extension
+    const fileExtension: string | undefined = fileName
+      .split('.')
+      .pop()
+      ?.toLowerCase()
+
+    if (fileExtension) {
+      if (!this.boolExtensionMap) {
+        // Populate the file extension map
+        // 1. Get Viya's response to this api call
+        const typesQueryUrl = `/types/types?limit=999999`
+        const response = (
+          await this.requestClient.get(typesQueryUrl, accessToken)
+        ).result as IViyaTypesResponse
+        // 2. Filter the returned items that have file extensions into a map
+        //    using forEach as an item may relate to multiple file extensions.
+        response.items
+          .filter((e) => e.extensions)
+          .forEach((e) => {
+            e.extensions?.forEach((ext) => {
+              this.fileExtensionMap.set(ext, {
+                // "name:" is the typeDefName value required for file creation.
+                typeDefName: e.name,
+                properties: e.properties
+              })
+            })
+          })
+        // 3. Toggle the flag to avoid repeating this step
+        this.boolExtensionMap = true
+      }
+
+      // Query the map for the current file extension
+      const fileExtInfo = this.fileExtensionMap.get(fileExtension)
+      if (fileExtInfo) {
+        // If the extension was found in the map, record the typeDefName and
+        // create a patch if a properties object was returned.
+        typeDefName = fileExtInfo.typeDefName
+        if (fileExtInfo.properties)
+          filePatch = { name: fileName, properties: fileExtInfo.properties }
+      }
+    }
+
+    // Create the file
+    const createFileResponse = await this.requestClient.post<File>(
+      `/files/files?parentFolderUri=${parentFolderUri}&typeDefName=${
+        typeDefName ?? 'file'
```

Comment:
When typeDefName is not available, it will be typeDefName= as an empty query parameter, is it just ignored in SAS or will it make troubles?